src/bin/pg_autoctl/pgsql.c

```diff
@@ -265,16 +264,49 @@ pgsql_compute_connection_retry_sleep_time(PGSQL *pgsql)
 	 * time spent, something we care to optimize for even when it means more
 	 * work on the monitor side.
 	 */
-	int previousSleepTime = pgsql->retryPolicy.sleepTime;
-	int sleepTime = random_between(pgsql->retryPolicy.baseSleepTime,
-								   previousSleepTime * 3);
+	int previousSleepTime = retryPolicy->sleepTime;
+	int sleepTime =
+		random_between(retryPolicy->baseSleepTime, previousSleepTime * 3);
 
-	pgsql->retryPolicy.sleepTime =
-		min(pgsql->retryPolicy.maxSleepTime, sleepTime);
+	retryPolicy->sleepTime = min(retryPolicy->maxSleepTime, sleepTime);
 
-	++(pgsql->retryPolicy.attempts);
+	++(retryPolicy->attempts);
 
-	return pgsql->retryPolicy.sleepTime;
+	return retryPolicy->sleepTime;
+}
+
+
+/*
+ * pgsql_retry_policy_expired returns true when we should stop retrying, either
+ * per the policy (maxR / maxT) or because we received a signal that we have to
+ * obey.
+ */
+bool
+pgsql_retry_policy_expired(ConnectionRetryPolicy *retryPolicy)
+{
+	uint64_t now = time(NULL);
+
+	/* Any signal is reason enough to break out from this retry loop. */
+	if (asked_to_quit || asked_to_stop || asked_to_stop_fast || asked_to_reload)
+	{
+		return true;
+	}
+
+	/*
+	 * We stop retrying as soon as we have spent all of our time budget or all
+	 * of our attempts count budget, whichever comes first.
+	 *
+	 * maxR = 0 (zero) means no retry at all, checked before the loop
+	 * maxR < 0 (zero) means unlimited number of retries
+	 */
+	if ((now - retryPolicy->startTime) >= retryPolicy->maxT ||
+		(retryPolicy->maxR > 0 &&
+		 retryPolicy->attempts >= retryPolicy->maxR))
+	{
+		return false;
```

Comment:
As also described in my other comment. I believe this should be `return true` and below `return false`.

Reply:
Done that way, you're right.